testbed/interfaces/node.go

```diff
@@ -0,0 +1,145 @@
+package testbedi
+
+import (
+	"context"
+	"io"
+)
+
+// NewNodeFunc constructs a node implementing the Core interface. It is provided
+// a path to an already created directory `dir`, as well as a map of attributes
+// which can be supplied to shape process execution.
+// Examples of attributes include: which binary to use, docker image, cpu/ram
+// limits, or any other information that may be required to property setup or
+// manage the node.
+type NewNodeFunc func(dir string, attrs map[string]interface{}) (Core, error)
+
+// GetAttrListFunc returns a list of attribute names that can be queried from
+// the node. These attributes may include those can be set from the NewNodeFunc,
+// or additional attributes at may only be available after initialization.
+// Attributes returned should be queriable through the Attribute interface.
+// Examples include: api address, peerid, cpu/ram limits, jitter.
+type GetAttrListFunc func() []string
+
+// GetAttrDescFunc returns the description of the attribute `attr`
+type GetAttrDescFunc func(attr string) (string, error)
+
+type Libp2p interface {
+	// PeerID returns the peer id
+	PeerID() (string, error)
+	// APIAddr returns the multiaddr for the api
+	APIAddr() (string, error)
+	// SwarmAddrs returns the swarm addrs for the node
+	SwarmAddrs() ([]string, error)
+}
+
+type Config interface {
+	Core
+	// GetConfig returns the configuration of the node
+	GetConfig() (interface{}, error)
+	// WriteConfig writes the configuration of the node
+	WriteConfig(interface{}) error
+}
+
+// Attributes are ways to shape process execution and additional information that alters the
+// environment the process executes in
+type Attribute interface {
```

Comment:
Is this an "Attribute" or "Attribute*s*"?

Reply:
It's just an extension of the Core interface, so not a singular or plural. The interfaces were split to reduce the required amount of code to implement a plugin. Example, if a node does not implement the `Attribute` interface, when running commands such as `iptb attr get`, an error will be returned stating that the operation can't be done because the plugin does not support it.